src/toolchain_override.rs

```diff
@@ -33,9 +36,96 @@ pub struct OverrideCfg {
 }
 
 // Represents the [toolchain] table in 'fuel-toolchain.toml'.
-#[derive(Debug, Deserialize, Serialize)]
+#[derive(Debug, Deserialize)]
 pub struct ToolchainCfg {
-    pub channel: String,
+    #[serde(deserialize_with = "deserialize_channel")]
+    pub channel: Channel,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Channel {
+    pub name: String,
+    pub date: Option<Date>,
+}
+
+impl Serialize for ToolchainCfg {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut c = serializer.serialize_struct("ToolchainCfg", 2)?;
+        c.serialize_field("channel", &self.channel.to_string())?;
+
+        c.end()
+    }
+}
+
+pub fn deserialize_channel<'de, D>(deserializer: D) -> Result<Channel, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let channel_str = String::deserialize(deserializer)?;
+
+    if is_beta_toolchain(&channel_str) {
+        return Ok(Channel {
+            name: channel_str,
+            date: None,
+        });
+    }
+
+    if let Some((name, date)) = channel_str.split_once('-') {
+        return Ok(Channel {
+            name: name.to_string(),
+            date: Date::parse(date, DATE_FORMAT)
+                .map_err(de::Error::custom)
+                .ok(),
+        });
+    }
+
+    if channel_str == LATEST || channel_str == NIGHTLY {
+        return Err(Error::invalid_value(
+            serde::de::Unexpected::Str(&channel_str),
+            &"channel with date",
+        ));
+    }
+
+    Err(Error::invalid_value(
+        serde::de::Unexpected::Str(&channel_str),
+        &"a valid channel str",
+    ))
+}
+
+impl fmt::Display for Channel {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.date {
+            Some(d) => write!(f, "{}-{}", self.name, d),
+            None => write!(f, "{}", self.name),
+        }
+    }
+}
+
+impl FromStr for Channel {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self> {
+        if is_beta_toolchain(s) {
+            return Ok(Self {
+                name: s.to_string(),
+                date: None,
+            });
+        };
+
+        if let Some((name, d)) = s.split_once('-') {
+            Ok(Self {
+                name: name.to_string(),
+                date: Date::parse(d, DATE_FORMAT).ok(),
+            })
+        } else {
+            if s == LATEST || s == NIGHTLY {
+                bail!("'{s}' without date specifier is forbidden");
+            }
+            bail!("Invalid str for channel: '{}'", s);
+        }
+    }
 }
```

Comment:
Can you re-use this `FromStr` implementation in the `deserialize_channel` function above? I.e. call `Channel::parse` within `deserialize_channel`? It looks like there's quite a bit of duplicated logic between the two.